evalml/pipelines/utils.py

```diff
@@ -369,6 +387,120 @@ def _make_new_component_name(model_type, component_name, idx=None):
+    """
+
+    def _make_new_component_name(name, component_name, idx=None, pipeline_name=None):
+        idx = " " + str(idx) if idx is not None else ""
+        if pipeline_name:
+            return f"{pipeline_name} Pipeline{idx} - {component_name}"
+        return f"{str(name)} Pipeline{idx} - {component_name}"
+
+    parameters = parameters if parameters else {}
+    final_components = []
+    used_names = []
+    component_graph = (
+        {"Label Encoder": ["Label Encoder", "X", "y"]}
+        if is_classification(problem_type)
+        else {}
+    )
+    for pipeline in input_pipelines:
+        component_pipeline_name = pipeline.name
+        name_idx = (
+            used_names.count(component_pipeline_name) + 1
+            if used_names.count(component_pipeline_name) > 0
+            else None
+        )
+        used_names.append(component_pipeline_name)
+        sub_pipeline_name = (
+            sub_pipeline_names[pipeline.name] if sub_pipeline_names else None
+        )
+        final_component = None
+        final_y = "y"
+        final_y_candidate = (
+            None
+            if not handle_component_class(
+                pipeline.component_graph.compute_order[-1]
+            ).modifies_target
+            else _make_new_component_name(
+                component_pipeline_name,
+                pipeline.component_graph.compute_order[-1],
+                name_idx,
+                sub_pipeline_name,
+            )
+            + ".y"
+        )
+        for name, component_list in pipeline.component_graph.component_dict.items():
+            new_component_list = []
+            new_component_name = _make_new_component_name(
+                component_pipeline_name, name, name_idx, sub_pipeline_name
+            )
+            for i, item in enumerate(component_list):
+                if i == 0:
+                    fitted_comp = handle_component_class(item)
+                    new_component_list.append(fitted_comp)
+                    parameters[new_component_name] = pipeline.parameters.get(name, {})
+                elif isinstance(item, str) and item not in ["X", "y"]:
+                    new_component_list.append(
+                        _make_new_component_name(
+                            component_pipeline_name, item, name_idx, sub_pipeline_name
+                        )
+                    )
+                elif isinstance(item, str) and item == "y":
+                    if is_classification(problem_type):
+                        new_component_list.append("Label Encoder.y")
+                    else:
+                        new_component_list.append("y")
+                else:
+                    new_component_list.append(item)
+                if i != 0 and item.endswith(".y"):
+                    final_y = _make_new_component_name(
+                        component_pipeline_name, item, name_idx, sub_pipeline_name
+                    )
```

Comment:
is it possible to move this check

```python
if i != 0 and item.endswith(".y"):
...
```
to fall under
```python
elif isinstance(item, str) and item not in ["X", "y"]:
    comp = _make_new_component_name(
                      component_pipeline_name, item, name_idx, sub_pipeline_name
                  )
    new_component_list.append(comp)
    if i != 0 and item.endswith(".y"):
        final_y = comp
elif ...
```
 This just reduces extra calls to `_make_new_component_name`, which could save time and memory for larger runs

Reply:
Another optimization we could do is enumerate over the `input_pipelines`, and only when we are on the last pipeline, check for and set `final_y`. This way, if we have a larger number of pipelines in `input_pipelines`, we won't need to keep setting `final_y` and overwriting it. 
Not a needed change, but would be nice to be more efficient with memory usage if we can!